### PR TITLE
Add line breaking support

### PR DESCRIPTION
**Pull request by paladine**
Primarily required by MMUD

Supports telnet updating as screen sizes are moved around.

### CODE REVIEW

File: MBBSEmu/Session/LineBreaker.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
+using MBBSEmu.Extensions;
+using MBBSEmu.HostProcess;
+using MBBSEmu.HostProcess.ExportedModules;
+using MBBSEmu.HostProcess.Structs;
+using MBBSEmu.Memory;
+using MBBSEmu.Module;
+using MBBSEmu.Server;
+using MBBSEmu.Session.Enums;
+using MBBSEmu.TextVariables;
+using System;
```
Comment (enusbaum) [inline MBBSEmu/Session/LineBreaker.cs:10]: All these `using` statements can go, the only import this needs is `System`

File: MBBSEmu/Session/LineBreaker.cs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+                                lineBufferToRawBuffer[lineBufferLength++] = rawBufferLength - 1;
+                                // overflow to the next line
+                                if (lineBufferLength > TerminalColumns)
+                                {
+                                    doLineBreak(b);
+                                }
+                                break;
+                        }
+                        break;
+                }
```
Comment (enusbaum) [inline MBBSEmu/Session/LineBreaker.cs:117]: Probably add a `default` case to throw an exception if it falls through for whatever reason

File: MBBSEmu/Session/LineBreaker.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        /// </summary>
+        private int rawBufferLength = 0;
+
+        /// <summary>
+        /// Breaks buffer into lines and calls SendToClientMethod afterwards.
+        /// </summary>
+        /// <param name="buffer">Raw output buffer going to a client</param>
+        public void SendBreakingIntoLines(byte[] buffer)
+        {
+            foreach(byte b in buffer)
```
Comment (enusbaum) [inline MBBSEmu/Session/LineBreaker.cs:80]: var

File: MBBSEmu/Session/LineBreaker.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            rawBufferLength = 0;
+        }
+
+        /// <summary>
+        /// Performs a line break
+        /// </summary>
+        /// <param name="b">The character following the full line</param>
+        private void doLineBreak(byte b)
+        {
+            if (Char.IsWhiteSpace((char) b)) {
```
Comment (enusbaum) [inline MBBSEmu/Session/LineBreaker.cs:131]: `char` vs `Char`

File: MBBSEmu/Session/LineBreaker.cs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                    lineBuffer[0] = b;
+                    lineBufferToRawBuffer[0] = rawBufferLength + 1;
+                    lineBufferLength = 1;
+
+                    rawBufferLength  += 2;
+                }
+                else
+                {
+                    // i is the location to place the \r\n
+                    int insertIndex = lineBufferToRawBuffer[lastSpaceIndex];
```
Comment (enusbaum) [inline MBBSEmu/Session/LineBreaker.cs:163]: var

File: MBBSEmu/Session/SessionBase.cs [hunk 6]
```diff
@@ -26,9 +26,17 @@ namespace MBBSEmu.Session
...
+
         protected delegate void SendToClientDelegate(byte[] dataToSend);
 
-        protected SendToClientDelegate SendToClientMethod;
+        private SendToClientDelegate _sendToClientMethod;
+
+        protected SendToClientDelegate SendToClientMethod
+        {
+            get => _sendToClientMethod;
+            set { _sendToClientMethod = value; _lineBreaker.SendToClientMethod = (buffer => value.Invoke(buffer)); }
```
Comment (enusbaum) [inline MBBSEmu/Session/SessionBase.cs:38]: Use a method group here:

`_lineBreaker.SendToClientMethod = value.Invoke;`

File: MBBSEmu/Session/Telnet/TelnetSession.cs [hunk 7]
```diff
@@ -183,5 +185,20 @@ private void OnIacVerbReceived(object sender, IacFilter.IacVerbReceivedEventArgs
                 base.Send(msIacToSend.ToArray());
             }
         }
+
+        private void OnIacSubnegotiationReceived(object sender, IacFilter.IacSubnegotiationEventArgs args)
+        {
+            if (args.Option == EnumIacOptions.NegotiateAboutWindowSize)
+            {
+                int columns = args.Data[0] << 8 | args.Data[1];
```
Comment (enusbaum) [inline MBBSEmu/Session/Telnet/TelnetSession.cs:193]: var all the things

**Review by enusbaum — CHANGES_REQUESTED** [review]

File: MBBSEmu/Session/LineBreaker.cs (hunk 2 above)
Comment (paladine) [reply MBBSEmu/Session/LineBreaker.cs:117]: default is intended, so just commented

**Pull request merged**